File: 018.4Sum/sum.cc
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        for (int i = 0; i < int(nums.size()) - 3; i++) {
            if (i != 0 && nums[i] == nums[i-1]) {
                continue;
            }
            for (int j = i + 1; j < int(nums.size()) - 2; j++) {
                if (j != i+1 && nums[j] == nums[j-1]) {
                    continue;
                }
                int l = j + 1, r = int(nums.size()) - 1;
                while (l < r) {
                    int sum = nums[i] + nums[j] + nums[l] + nums[r];
                    if (sum == target) {
                        vector<int> tmp = {nums[i], nums[j], nums[l], nums[r]};
                        ans.push_back(tmp);
                        while (l < r && nums[l] == nums[l+1]) l++;
                        while (l < r && nums[r] == nums[r-1]) r--;
                        l++;
                        r--;
                    }
                    else if (sum < target) l++;
                    else r--;
                }
            }
        }
        return ans;
    }
};

File: 018.4Sum/sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        n = len(nums)
        ans = []

        for i in range(n-3):
            if i != 0 and nums[i] == nums[i-1]:
                continue

            for j in range(i+1, n-2):
                if j != i+1 and nums[j] == nums[j-1]:
                    continue
                l, r = j+1, n-1

                while l < r:
                    s = nums[i] + nums[j] + nums[l] + nums[r]
                    if s == target:
                        ans.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l+1]: l += 1
                        while l < r and nums[r] == nums[r-1]: r -= 1
                        l += 1
                        r -= 1

                    elif s < target: l += 1
                    else: r -= 1

        return ans

if __name__ == '__main__':
    nums = [-1,0,-5,-2,-2,-4,0,1,-2]
    target = -9
    print Solution().fourSum(nums, target)
